tests/Message/PurchaseRequestTest.php

```diff
@@ -75,4 +76,115 @@ public function testSetConsumerLanguage()
         $this->request->setConsumerLanguage('en');
         $this->assertSame('002', $this->request->getConsumerLanguage());
     }
+
+    public function test3DsEmvFields()
```

Comment:
This unit test is a "test everything, all at once" - while useful, it assumes the request is being modified (rather than initialized), and doesn't provide coverage for failure cases, malformed input, or a minimal initialization that is still a valid 3DS transaction. There are no additional tests/differences to the gateway tests either - while not necessarily a requirement, it's a good option to cover in the unit tests to ensure that the request is being parsed correctly at the Redsys end and returning the appropriate response. i.e. how do they respond when the 3DS fields aren't appropriate, or not enough of them, etc.

Reply:
Ok removed this test and added tests for each component, getters, setters, nulls, invalid data, etc.

Reply:
Lots of request tests, which is great - good to add gateway tests as well so we can check the responses coming back from Redsys are matching expectations.